File: crypto/test/test_puid_state_record_serializer.cpp
#include <library/cpp/testing/gtest/gtest.h>

#include <crypta/styx/services/common/data/proto/puid_state.pb.h>
#include <crypta/styx/services/common/data/proto_comparators.h>
#include <crypta/styx/services/common/data/test_helpers.h>
#include <crypta/styx/services/common/serializers/puid_state_record_serializer.h>

#include <util/generic/vector.h>

#include <utility>

using namespace NCrypta::NStyx;
using namespace NCrypta::NStyx::NTest;

namespace {
    const ui64 REF_PUID = 100500;
    const TVector<TOblivionEvent> REF_OBLIVION_EVENTS = {
            CreateOblivionEvent(TInstant::Seconds(1600000000), "some_hash"),
            CreateOblivionEvent(TInstant::Seconds(1700000000), "some_new_hash"),
    };
    const TString REF_SERIALIZED_KEY = "puid:100500";
    const TString REF_SERIALIZED_VALUE = "\x12\x11\x08\x80\xA0\xF8\xFA\5\x12\tsome_hash\x12\x15\x08\x80\xE2\xCF\xAA\6\x12\rsome_new_hash";
}

TEST(NPuidStateRecordSerializer, Serialize) {
    const auto& puidState = NTest::GetPuidState(REF_PUID, REF_OBLIVION_EVENTS);

    const auto& record = NPuidStateRecordSerializer::ToRecord(puidState);

    EXPECT_EQ(REF_SERIALIZED_KEY, record.Key);
    EXPECT_EQ(REF_SERIALIZED_VALUE, record.Value);
}

TEST(NPuidStateRecordSerializer, Deserialize) {
    const NCrypta::TRecord record = {.Key= REF_SERIALIZED_KEY, .Value = REF_SERIALIZED_VALUE};

    const auto puidState = NPuidStateRecordSerializer::FromRecord(record);

    EXPECT_EQ(REF_PUID, puidState.GetPuid());
    EXPECT_EQ(REF_OBLIVION_EVENTS, OblivionEventsToVector(puidState.GetOblivionEvents()));
}

File: crypto/test/test_puid_state_record_serializer.py
import pytest

from crypta.lib.native.database.python.record import TRecord
from crypta.styx.services.common.data.proto.puid_state_pb2 import (
    TOblivionEvent,
    TPuidState,
)
from crypta.styx.services.common.serializers.python import puid_state_record_serializer


REF_KEY = "puid:100500"
REF_VALUE = b"\x12\t\x08\x94\x91\x06\x12\x03foo\x12\t\x08\xa4\xdf\x06\x12\x03bar"


@pytest.fixture(scope="function")
def ref_puid_state():
    state = TPuidState()

    state.Puid = 100500
    state.OblivionEvents.append(TOblivionEvent(Unixtime=100500, Obfuscated="foo"))
    state.OblivionEvents.append(TOblivionEvent(Unixtime=110500, Obfuscated="bar"))

    return state


def test_to_record(ref_puid_state):
    record = puid_state_record_serializer.ToRecord(ref_puid_state)
    assert REF_KEY == record.Key
    assert REF_VALUE == record.Value


def test_from_record(ref_puid_state):
    record = TRecord.Create(REF_KEY, REF_VALUE)
    assert ref_puid_state == puid_state_record_serializer.FromRecord(record)
